File: bindings/python/perf/single/perf_router_router.py
import sys
import threading
import time

import zlink

from perf_common import (
    STOP_TOKEN,
    apply_single_socket_options,
    benchmark_run_id,
    configure_single_tls_client,
    configure_single_tls_server,
    new_payload,
    parse_single_args,
    perf_context,
    poll_idle_ms,
    print_result_lines,
    new_single_latency_sampler,
    run_one_way_receiver,
    result_metrics,
    resolve_single_endpoint,
    resolve_single_connect_ready_timeout_ms,
    single_routing_probe,
    send_routed_sync,
    stamp_payload,
    wait_monitor_event,
)


def _send_router_stop_token(router, dest_routing_id):
    """PERF_SINGLE_TEST_POLICY § 1.4 wire-level shutdown signal.

    Router-router uses the synchronous public send terminal; the stop token is a
    single payload frame addressed to the peer.
    """

    for _ in range(100):
        try:
            router.send(dest_routing_id).message(STOP_TOKEN).submit_sync(
                flags=zlink.SendFlags.NONE
            )
            return
        except zlink.SubmitError as exc:
            if exc.result not in (
                zlink.SubmitResult.BACKPRESSURED,
                zlink.SubmitResult.NOT_CONNECTED,
            ):
                raise
            poll_idle_ms(1)


def _public_one_way_metrics(sender, receiver, *, msg_size, duration_s, run_id):
    return None


def main(argv=None):
    args = parse_single_args(argv or sys.argv[1:], pattern="router_router")
    run_id = benchmark_run_id()
    latency_sampler = new_single_latency_sampler()
    received = 0
    payload = new_payload(args.msg_size)

    def send_loop(router, active_end):
        # Core owns blocking HWM admission on this dedicated sender thread.
        stamp = stamp_payload
        while time.perf_counter() < active_end:
            send_routed_sync(
                router,
                stamp(payload, phase=1, run_id=run_id),
                routing_id=b"SERVER",
            )
        _send_router_stop_token(router, b"SERVER")

    with perf_context() as ctx:
        with zlink.create_router_socket(ctx) as server:
            with zlink.create_router_socket(ctx) as client:
                server.set_routing_id(b"SERVER")
                client.set_routing_id(b"CLIENT")
                client.router_options.connect_routing_id = b"SERVER"
                endpoint = resolve_single_endpoint(args.transport, "router-router")
                apply_single_socket_options(server, client)
                configure_single_tls_server(server, args.transport)
                configure_single_tls_client(client, args.transport)
                with client.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as monitor:
                    server.bind(endpoint)
                    client.connect(endpoint)
                    wait_monitor_event(
                        monitor,
                        zlink.MonitorEventMask.CONNECTION_READY,
                        timeout_ms=resolve_single_connect_ready_timeout_ms(),
                    )

                # C perf_router_router.cpp wait_for_router_router_ready:
                # one-shot routed probe (addressed to the peer routing id)
                # before phase=active.
                if not single_routing_probe(
                    client,
                    server,
                    payload,
                    run_id=run_id,
                    msg_size=args.msg_size,
                    routing_id=b"SERVER",
                ):
                    raise RuntimeError(
                        "router-router routing probe did not establish route"
                    )

                active_end = time.perf_counter() + args.duration
                metrics = _public_one_way_metrics(
                    client,
                    server,
                    msg_size=args.msg_size,
                    duration_s=args.duration,
                    run_id=run_id,
                )
                if metrics is None:
                    sender = threading.Thread(
                        target=send_loop, args=(client, active_end), daemon=True
                    )
                    sender.start()
                    # C perf_router_router.cpp run_active_phase receiver.
                    received = run_one_way_receiver(
                        server,
                        method="recv",
                        msg_size=args.msg_size,
                        run_id=run_id,
                        active_end=active_end,
                        received=received,
                        latency_sampler=latency_sampler,
                    )

                    sender.join()
                    if received == 0:
                        raise RuntimeError(
                            "router-router benchmark did not receive any active message"
                        )
                    metrics = result_metrics(
                        count=received,
                        msg_size=args.msg_size,
                        elapsed_s=args.duration,
                        latency_sampler=latency_sampler,
                    )
                print_result_lines("ROUTER_ROUTER", args.transport, args.msg_size, metrics)


if __name__ == "__main__":
    main()

File: bindings/node/perf/single/perf_router_router.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const zlink = require('@zlink-systems/zlink');
const {
  createMetricCollector,
  createRunId,
  currentEpochNs,
  integerEnv,
  summarizeMetrics,
} = require('../common/perf_metrics');
const {
  applyContextPolicy,
  applySocketPolicy,
  appendMeasurement,
  benchmarkEndpoint,
  closeSenderWorker,
  configureTlsServer,
  drainRouterRecvInto,
  emitSingleSocketHwmDetail,
  parseSingleBinaryArgs,
  runLocalSocketOneWayBenchmark,
  spawnSenderWorker,
  waitForWorkerError,
  waitForWorkerMessage,
} = require('./perf_single_common');
const { STOP_TOKEN_BYTES } = require('../perf_stop_token');

const RECEIVER_ID = Buffer.from('ROUTER1', 'ascii');
const SENDER_ID = Buffer.from('ROUTER2', 'ascii');
const RECEIVER_ROUTING_ID = zlink.RoutingId.from(RECEIVER_ID);

function trace(message) {
  if (process.env.PERF_NODE_TRACE === '1') {
    console.error(`[router-router] ${message}`);
  }
}

function partStrings(received) {
  return received.parts.map((part) => part.data().toString());
}

async function handshakeRouterReceiver(receiver) {
  const ping = new zlink.Received();
  receiver.recv(ping);
  try {
    if (ping.routingId === null || partStrings(ping).join(',') !== 'PING') {
      throw new Error('router-router handshake receive failed');
    }

    await receiver.send(ping.routingId).message(Buffer.from('PONG')).submit();
    return ping.routingId;
  } finally {
    ping.close();
  }
}

async function handshakeRouterReceiverWithRetry(receiver) {
  const ping = new zlink.Received();
  try {
    const configuredMs = integerEnv('PERF_ROUTER_HANDSHAKE_TIMEOUT_MS', 3000);
    const timeoutMs = configuredMs > 0 ? configuredMs : 3000;
    const deadlineNs = process.hrtime.bigint() + BigInt(timeoutMs) * 1_000_000n;
    let received = false;
    while (!received && process.hrtime.bigint() < deadlineNs) {
      try {
        received = receiver.recv(ping, zlink.RecvFlags.DontWait);
      } catch (error) {
        if (!(error instanceof zlink.RecvError && error.result === zlink.RecvResult.NoData)) {
          throw error;
        }
      }
      if (!received) {
        await new Promise((resolve) => setTimeout(resolve, 1));
      }
    }
    if (!received) {
      throw new Error('router-router handshake ping timeout');
    }
    if (ping.routingId === null || partStrings(ping).join(',') !== 'PING') {
      throw new Error('router-router handshake receive failed');
    }

    await receiver.send(ping.routingId).message(Buffer.from('PONG')).submit();
    return ping.routingId;
  } finally {
    ping.close();
  }
}

async function runRouterRouterBenchmark(msgSize, options) {
  if (options.transport === 'inproc') {
    // inproc is context-local so the Worker sender path cannot reach it.
    // Run both ROUTER sockets in one shared context with the C-faithful
    // blocking-equivalent one-way model + the PING/PONG routing-id gate
    // (C perf_router_router.cpp does the same handshake before active).
    return runLocalSocketOneWayBenchmark({
      pattern: 'ROUTER_ROUTER',
      msgSize,
      options,
      endpointToken: 'router-router',
      createReceiver: (ctx) => zlink.createRouterSocket(ctx),
      createSender: (ctx) => zlink.createRouterSocket(ctx),
      configureReceiver: (socket) => socket.setRoutingId(RECEIVER_ROUTING_ID),
      configureSender: (socket) => socket.setRoutingId(zlink.RoutingId.from(SENDER_ID)),
      handshake: async (sender, receiver) => {
        await sender.send(RECEIVER_ROUTING_ID)
          .message(Buffer.from('PING')).submit();
        const senderRid = await handshakeRouterReceiver(receiver);
        const reply = new zlink.Received();
        sender.recv(reply);
        try {
          if (partStrings(reply).join(',') !== 'PONG') {
            throw new Error('router-router handshake reply failed');
          }
        } finally {
          reply.close();
        }
        // Receiver replies/active are addressed by the sender's routing
        // id; the sender addresses the receiver by RECEIVER_ROUTING_ID.
        return RECEIVER_ROUTING_ID;
      },
      sendActive: async (socket, payload, routingId) => {
        try {
          await appendMeasurement(socket.send(routingId), payload).submit();
          return true;
        } catch (error) {
          if (error instanceof zlink.SubmitError
            && (error.result === zlink.SubmitResult.Backpressured
              || error.result === zlink.SubmitResult.NotConnected
              || error.result === zlink.SubmitResult.NotFound)) {
            return false;
          }
          const text = String(error && error.message ? error.message : error);
          if ((error && error.code === 'EAGAIN')
            || text.includes('Resource temporarily unavailable')) {
            return false;
          }
          throw error;
        }
      },
      sendStop: async (socket, routingId) => {
        await socket.send(routingId).message(STOP_TOKEN_BYTES).submit();
      },
    });
  }

  const ctx = zlink.createContext();
  applyContextPolicy(ctx);
  const receiver = zlink.createRouterSocket(ctx);
  const endpoint = await benchmarkEndpoint(options.transport, `router-router-${msgSize}`);
  let worker = null;

  try {
    applySocketPolicy(receiver, options);
    ctx.recalculateAutoHwm();
    receiver.setRoutingId(RECEIVER_ROUTING_ID);
    configureTlsServer(receiver, options.transport);
    receiver.bind(endpoint);
    worker = spawnSenderWorker({
      kind: 'router_router',
      transport: options.transport,
      endpoint,
      duration: options.duration,
      msgSize,
      runId: options.runId ?? 1,
      receiverRoutingIdBytes: RECEIVER_ID,
      senderRoutingIdBytes: SENDER_ID,
      options,
    });
    const workerError = waitForWorkerError(worker);
    await Promise.race([
      waitForWorkerMessage(worker, 'connected'),
      workerError.then((message) => Promise.reject(new Error(message.message)))
    ]);
    worker.postMessage({ type: 'handshake' });
    await handshakeRouterReceiverWithRetry(receiver);
    await Promise.race([
      waitForWorkerMessage(worker, 'ready'),
      workerError.then((message) => Promise.reject(new Error(message.message)))
    ]);
    trace('handshake done');

    const activeStartNs = currentEpochNs();
    const activeStopNs = activeStartNs
      + BigInt(Math.floor(options.duration * 1_000_000_000));
    const runId = createRunId(options.runId ?? 1);
    const collector = createMetricCollector({
      runId,
      msgSize,
      activeStartNs,
      activeStopNs,
    });

    // PERF_SINGLE_TEST_POLICY § 1.4 / § 2.0.1: the PING/PONG handshake
    // above is the routing-id discovery gate (C perf_router_router.cpp
    // does the same). No extra start/stop control channel — the receiver
    // uses blocking recv + drain and exits on the wire stop token.
    const recvTask = drainRouterRecvInto(
      receiver,
      msgSize,
      Object.assign(collector, { runId, activeStartNs }),
      { recordUntilNs: activeStopNs }
    );
    await Promise.race([
      recvTask,
      workerError.then((message) => Promise.reject(new Error(message.message)))
    ]);
    const result = collector.finish();
    emitSingleSocketHwmDetail(receiver, 'ROUTER_ROUTER', options.transport, 'receiver', msgSize);
    return result;
  } finally {
    trace('closing');
    await closeSenderWorker(worker);
    receiver.close();
    trace('receiver closed');
    ctx.close();
    trace('ctx closed');
  }
}

module.exports = { runRouterRouterBenchmark };

if (require.main === module) {
  (async () => {
    const options = parseSingleBinaryArgs(process.argv.slice(2));
    const result = await runRouterRouterBenchmark(options.msgSize, options);
    if (result.unsupported) {
      console.log(`UNSUPPORTED,${options.libName},ROUTER_ROUTER,${options.transport}`);
      return;
    }
    for (const line of summarizeMetrics(
      'ROUTER_ROUTER',
      options.transport,
      options.msgSize,
      result.latenciesNs,
      options.duration,
      options.libName,
      result.accepted,
      result.latencyMeanNs
    )) {
      console.log(line);
    }
  })().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}
